{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 2: Primer to NGS Data Analysis\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "\n",
    "- [1. Introduction to NGS Data Analysis](#1.-Introduction) \n",
    "- [2. Genome and Simulated Read Data](#2.-Genome-and-Simulated-Read-Data)\n",
    "- [3. Short read alignment](#3.-Read-alignment)\n",
    "- [4. samtools:-Manipulating-the-alignment-files](#4.-samtools:-Manipulating-the-alignment-files)\n",
    "- [5. Understanding the VCF Files](#5.-Understanding-the-VCF-Format)\n",
    "- [6. Visualizing the alignment](#6.-Visualing-the-mapping-of-reads)\n",
    "- [7. Further reading](#7.-References)\n",
    "- [8. Exercises]\n",
    "\n",
    "### Prerequisite installation\n",
    "\n",
    "1. [wgsim](https://github.com/lh3/wgsim): Reads simulator.\n",
    "2. [bwa](https://github.com/lh3/bwa): short-read alignment tools.\n",
    "3. [samtools](https://samtools.sourceforge.net): SAM manipulating tools.\n",
    "4. [bcftools](https://github.com/samtools/bcftools): VCF file manipulator."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Introduction\n",
    "\n",
    "Next-generation sequencing (NGS) refers to the emerging high-throughput technologies for sequencing DNA or RNA. A typical workflow for NGS data analysis is usually composed of the following steps:\n",
    "\n",
    "![](../images/ngs_overview.png)\n",
    "\n",
    "`bwa` and `samtools` are open-source tools for NGS data analyis, which fit into step3, step 4-5 respectively.\n",
    "\n",
    "[BWA](http://bio-bwa.sourceforge.net/) is probably the most widely used short read aligner ever. By now, it’s even part of the official Illumina analysis pipeline. BWA actually contains three different aligner algorithms:\n",
    "* BWA-Backtrack: used for short-read alignment using FM-Index strategy.\n",
    "* BWA-SW: used for long reads alignment, utilizing seed (FM-index) and extend (smith-waterman) strategy.\n",
    "* BWA-Mem: used for long-read alignment.\n",
    "\n",
    "\n",
    "As a companion tools, `samtools` can process aligned sequence reads, and manipulate them with ease:\n",
    "* convert between the two most common file formats (SAM and BAM), \n",
    "* sort and index files (for speedy retrieval later),\n",
    "* extract specific genomic regions of interest.\n",
    "\n",
    "`samtools` also enables quality checking of reads, and automatic identification of genomic variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Genome and Simulated Read Data\n",
    "\n",
    "For simplicity, we take a small set of simulated reads from *E. coli*. We use E. coli because its genome is quite small — 4,649 kilobases, with 4,405 genes, and its entire genome fits into a single FASTA file of 4.8 megabytes.\n",
    "\n",
    "### 2.1 Genome files\n",
    "\n",
    "You can download directly from NCBI ftp to get the genome sequence file for E.coli.\n",
    "\n",
    "### 2.2 Generating simulated reads\n",
    "\n",
    "Here we use `[wgsim](https://github.com/lh3/wgsim)` tools to generate a small set of artificial reads data.\n",
    "```bash\n",
    "Program: wgsim (short read simulator)\n",
    "Version: 0.3.0\n",
    "Contact: Heng Li <lh3@sanger.ac.uk>\n",
    "\n",
    "Usage:   wgsim [options] <in.ref.fa> <out.read1.fq> <out.read2.fq>\n",
    "\n",
    "Options: -e FLOAT      base error rate [0.020]\n",
    "         -d INT        outer distance between the two ends [500]\n",
    "         -s INT        standard deviation [50]\n",
    "         -N INT        number of read pairs [1000000]\n",
    "         -1 INT        length of the first read [70]\n",
    "         -2 INT        length of the second read [70]\n",
    "         -r FLOAT      rate of mutations [0.0010]\n",
    "         -R FLOAT      fraction of indels [0.15]\n",
    "         -X FLOAT      probability an indel is extended [0.30]\n",
    "         -S INT        seed for random generator [-1]\n",
    "         -h            haplotype mode\n",
    "```\n",
    "\n",
    "By default, `wgsim` will simulate $10^6$ paired-end reads. Here you need to generate $10^4$ single-end reads of length 50 by setting `out.read2.fq` to be `/dev/null`, while setting point mutation rate, indels fraction and the other options as default:\n",
    "```bash\n",
    "mkdir simulated_reads\n",
    "wgsim -N 10000 -d 300 -1 50 -2 50 -S 1234 genomes/NC_008253.fna simulated_reads/sim_reads.fq /dev/null\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Read alignment\n",
    "\n",
    "The next step is to align the artificial reads to the reference genome. Here we use `bwa` to perform the task.\n",
    "\n",
    "### 3.1 Building genome index\n",
    "\n",
    "The first step for read alignment is to build the FM-index for the genome string. For `bwa`, the command is `bwa index`:\n",
    "```bash\n",
    "Usage:   bwa index [options] <in.fasta>\n",
    "\n",
    "Options: -a STR    BWT construction algorithm: bwtsw or is [auto]\n",
    "         -p STR    prefix of the index [same as fasta name]\n",
    "         -b INT    block size for the bwtsw algorithm (effective with -a bwtsw) [10000000]\n",
    "         -6        index files named as <in.fasta>.64.* instead of <in.fasta>.* \n",
    "\n",
    "Warning: `-a bwtsw' does not work for short genomes, while `-a is' and\n",
    "         `-a div' do not work not for long genomes.\n",
    "\n",
    "```\n",
    "\n",
    "Since E.coli is a small genome, we need to choose the `is` algorithm to build the genome index:\n",
    "```bash\n",
    "mkdir indexes\n",
    "bwa index -a is -p indexes/e_coli genomes/NC_008253.fna\n",
    "```\n",
    "This command will generate 5 files under the `indexes` directory:\n",
    "* `e_coli.amb`: summary of the ambiguous character in the genome sequence\n",
    "* `e_coli.ann`: annotation of the genome sequence\n",
    "* `e_coli.bwt`: Burrow-wheeler transform (BWT) of the genome\n",
    "* `e_coli.pac`: packed format of the genome string\n",
    "* `e_coli.sa`: sampled suffix array"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Align the short reads to the genome index\n",
    "\n",
    "`bwa` can perform 3 algorithms for read mapping：\n",
    "- `bwa-backtrack`: used for alignment of short reads shorter than 70bp;\n",
    "- `bwa-sw`: hybrid of Smith-Waterman algorithm and FM-index, for alignment of longer reads;\n",
    "- `bwa-mem`: memory-efficient algorithm for alignment of longer reads.\n",
    "\n",
    "Here we use `bwa-backtrack` algorithm (`bwa aln` + `bwa samse`) to conduct the alignment. Here is the help information for `bwa aln`:\n",
    "```bash\n",
    "Usage:   bwa aln [options] <prefix> <in.fq>\n",
    "\n",
    "Options: -n NUM    max #diff (int) or missing prob under 0.02 err rate (float) [0.04]\n",
    "         -o INT    maximum number or fraction of gap opens [1]\n",
    "         -e INT    maximum number of gap extensions, -1 for disabling long gaps [-1]\n",
    "         -i INT    do not put an indel within INT bp towards the ends [5]\n",
    "         -d INT    maximum occurrences for extending a long deletion [10]\n",
    "         -l INT    seed length [32]\n",
    "         -k INT    maximum differences in the seed [2]\n",
    "         -m INT    maximum entries in the queue [2000000]\n",
    "         -t INT    number of threads [1]\n",
    "         -M INT    mismatch penalty [3]\n",
    "         -O INT    gap open penalty [11]\n",
    "         -E INT    gap extension penalty [4]\n",
    "         -R INT    stop searching when there are >INT equally best hits [30]\n",
    "         -q INT    quality threshold for read trimming down to 35bp [0]\n",
    "         -f FILE   file to write output to instead of stdout\n",
    "         -B INT    length of barcode\n",
    "         -L        log-scaled gap penalty for long deletions\n",
    "         -N        non-iterative mode: search for all n-difference hits (slooow)\n",
    "         -I        the input is in the Illumina 1.3+ FASTQ-like format\n",
    "         -b        the input read file is in the BAM format\n",
    "         -0        use single-end reads only (effective with -b)\n",
    "         -1        use the 1st read in a pair (effective with -b)\n",
    "         -2        use the 2nd read in a pair (effective with -b)\n",
    "         -Y        filter Casava-filtered sequences\n",
    "```\n",
    "and `bwa samse` for single-end reads:\n",
    "```bash\n",
    "bwa samse [-n max_occ] [-f out.sam] [-r RG_line] <prefix> <in.sai> <in.fq>\n",
    "```\n",
    "\n",
    "Therefore, we run\n",
    "```bash\n",
    "bwa aln indexes/e_coli simulated_reads/sim_reads.fq -f alignments/sim_reads_mapped.sai\n",
    "bwa samse -f alignments/sim_reads_mapped.sam indexes/e_coli alignments/sim_reads_mapped.sai simulated_reads/sim_reads.fq\n",
    "```\n",
    "\n",
    "Have a look at the **sam index** file `*.sai` and the SAM file `*.sam`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Interpret the SAM output\n",
    "\n",
    "![](../images/samtools1.png)\n",
    "As SAMtools is primarily concerned with manipulating SAM files, it is useful to take a look at the sample SAM file generated by `bwa`, and to dive into the details of the SAM file format.\n",
    "\n",
    "A typical SAM file contains two types of lines:\n",
    "- **header** lines starting with a `@`, which provide the meta-data required for the alignment;\n",
    "- **alignment** lines starting without `@`, each line of which describe the single alignment information of each read against the reference genome. \n",
    "\n",
    "Let's start with the first alignment line for illustrative purpose, which contains 11 mandatory fields, followed by a number of user-defined optional fields.\n",
    "\n",
    "#### (1) Mandatory fields\n",
    "\n",
    "The mandatory fields include the following information:\n",
    "\n",
    "| Col | Field | Type | Description | Example |\n",
    "| --- | --- | --- | --- | --- |\n",
    "| 1 |  QNAME | str | query name of the read or the read pair | `NC_008253.1_0` |\n",
    "| 2 | FLAG | int | bitwise flags (pairing, mapped, mate mapped, etc.) | 16: indicates that the read maps to the reverse strand of the genome. | \n",
    "| 3 | RNAME | str | reference sequence name |  `NC_008253.1` |\n",
    "| 4 | POS | int | 1-based leftmost position of clipped alignment | `407` |\n",
    "| 5 | MAPQ | int | PHRED-scale mapping quality | `37` |\n",
    "| 6 | CIGAR | str | extended CIGAR string for alignment details | `70M` |\n",
    "| 7 | RNEXT | str | mate reference name ('=' for the same) |  * |\n",
    "| 8 | PNEXT | int | position of mate/next segment | 0 |\n",
    "| 9 | TLEN | int | observed template length | 0 |\n",
    "| 10 | SEQ | str | segment sequence | `CCAGGCAGTGGCAGGTGGCC...T` |\n",
    "| 11 | QUAL | str | ASCII of PHRED-scale base quality | `22222222222222222222...2` | \n",
    "\n",
    "#### (2) `FLAG` field\n",
    "Each bits in the `FLAG` field is defined as\n",
    "\n",
    "| Chr |\tFlag |\tDescription |\n",
    "| --- | --- | --- |\n",
    "| p | `0x0001` | the read is paired in sequencing |\n",
    "| P\t| `0x0002` | the read is mapped in a proper pair |\n",
    "| u\t| `0x0004` | the query sequence itself is unmapped |\n",
    "| U | `0x0008` |  the mate is unmapped |\n",
    "| r | `0x0010` |  strand of the query (1 for reverse) |\n",
    "| R | `0x0020` | strand of the mate |\n",
    "| 1 | `0x0040` | the read is the first read in a pair |\n",
    "| 2 | `0x0080`| the read is the second read in a pair |\n",
    "| s\t| `0x0100` | the alignment is not primary |\n",
    "| f | `0x0200` | QC failure |\n",
    "| d | `0x0400`|\toptical or PCR duplicate |\n",
    "\n",
    "\n",
    "#### (3) Optional fields\n",
    "\n",
    "The optional SAM fields always use the format: **FIELD_NAME:DATA_TYPE:VALUE**. Here is a list of  the possible **DATA_TYPEs**\n",
    "- `A`: **character**\n",
    "- `f`: **float**\n",
    "- `H`: **hexadecimal string**\n",
    "- `i`: **integer**\n",
    "- `Z`: **string**\n",
    "\n",
    "\n",
    "| Fields | Description | Data Type |\n",
    "| --- | --- |  --- |\n",
    "| `XT:A:U` |   The mapping `Type` is `Unique`        |  Character: Unique/Repeat/N/Mate-sw | \n",
    "| `NM:i:1` | \tThe edit distance to the reference is 1 | Integer | \n",
    "| `X0:i:1` |  Number of best-hit is 1 | Integer |\n",
    "|`X1:i:0` |\t Number of sub-optimal hits is 0 | Integer |\n",
    "| `XM:i:1` |  Number of the mismatches in the alignment is 1 | Integer |\n",
    "|`XO:i:0` |\t Number of gap-Open is 0 | Integer | \n",
    "| `XG:i:0` | Number of Gap-extensions is 0 | Integer | \n",
    "| `MD:Z:8G61` |  8 matches followed by a mismatch to `G` in reference and then 61 matches | String: Mismatch positions and bases |\n",
    "\n",
    "\n",
    "For further about SAM/BAM format, please refer to https://samtools.github.io/hts-specs/SAMv1.pdf. \n",
    "\n",
    "For FLAG information, you can refer to http://picard.sourceforge.net/explain-flags.html.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. `samtools`: Manipulating the alignment files\n",
    "\n",
    "Now that you understand how to conduct alignments, and know what each SAM field denotes. With the `SAM` files in hand, we could call variants out of the alignment results using `samtools`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1 Converting SAM to BAM\n",
    "\n",
    "To achieve the goal of variant calling, our first step is to convert the SAM file to BAM. This is an important prerequisite, as all the downstream steps, including the identification of genomic variants and visualization of reads, require BAM as input.\n",
    "\n",
    "This task, can be done using `samtools view`:\n",
    "```bash\n",
    "Usage:   samtools view [options] <in.bam>|<in.sam> [region1 [...]]\n",
    "\n",
    "Options: \n",
    "         -b       output BAM\n",
    "         -h       print header for the SAM output\n",
    "         -H       print header only (no alignments)\n",
    "         -S       input is SAM\n",
    "         -u       uncompressed BAM output (force -b)\n",
    "         -1       fast compression (force -b)\n",
    "         -x       output FLAG in HEX (samtools-C specific)\n",
    "         -X       output FLAG in string (samtools-C specific)\n",
    "         -c       print only the count of matching records\n",
    "         -B       collapse the backward CIGAR operation\n",
    "         -@ INT   number of BAM compression threads [0]\n",
    "         -L FILE  output alignments overlapping the input BED FILE [null]\n",
    "         -t FILE  list of reference names and lengths (force -S) [null]\n",
    "         -T FILE  reference sequence file (force -S) [null]\n",
    "         -o FILE  output file name [stdout]\n",
    "         -R FILE  list of read groups to be outputted [null]\n",
    "         -f INT   required flag, 0 for unset [0]\n",
    "         -F INT   filtering flag, 0 for unset [0]\n",
    "         -q INT   minimum mapping quality [0]\n",
    "         -l STR   only output reads in library STR [null]\n",
    "         -r STR   only output reads in read group STR [null]\n",
    "         -s FLOAT fraction of templates to subsample; integer part as seed [-1]\n",
    "         -?       longer help\n",
    "```\n",
    "\n",
    "So you can run\n",
    "```bash\n",
    "samtools view -b -S alignments/sim_reads_mapped.sam -o alignments/sim_reads_mapped.bam\n",
    "```\n",
    "\n",
    "### <font color=\"red\">$\\S$Exercise 1</font>\n",
    "\n",
    "1. How many of the alignments pass the mapping-quality threshold of 30? That is, the probability of mapping error is lower than $p=10^{-30/10} = 0.001$.\n",
    "\n",
    "2. How many reads fail to map to the reference genome, which has a `FLAG` of `0x0004`? Can you list the name of these reads?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2 Sorting and Indexing\n",
    "\n",
    "This is the help manual for `samtools sort`:\n",
    "```bash\n",
    "Usage:   samtools sort [options] <in.bam> <out.prefix>\n",
    "\n",
    "Options: \n",
    "         -n        sort by read name\n",
    "         -f        use <out.prefix> as full file name instead of prefix\n",
    "         -o        final output to stdout\n",
    "         -l INT    compression level, from 0 to 9 [-1]\n",
    "         -@ INT    number of sorting and compression threads [1]\n",
    "         -m INT    max memory per thread; suffix K/M/G recognized [768M]\n",
    "```\n",
    "and `samtools index`:\n",
    "```bash\n",
    "Usage: samtools index <in.bam> [out.index]\n",
    "```\n",
    "\n",
    "The next step is to sort and index the BAM file. There are two options for sorting BAM files: **by read name (-n)**, and **by genomic location (default)**. As our goal is to call genomic variants, and this requires that we \"pile-up\" all matching reads within a specific genomic location, we sort by location:\n",
    "```bash\n",
    "samtools sort alignments/sim_reads_aligned.bam alignments/sim_reads_aligned.sorted\n",
    "```\n",
    "\n",
    "Once you have sorted your BAM file, you can then index it. This enables tools, including SAMtools itself, and other genomic viewers to perform efficient random access on the BAM file, resulting in greatly improved performance. To do so, run:\n",
    "```bash\n",
    "samtools index alignments/sim_reads_aligned.bam\n",
    "```\n",
    "This will create an index file named `alignments/sim_reads_aligned.bam.bai`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.3 Identifying genomic variants\n",
    "\n",
    "Now, we are ready to identify the genomic variants from our reads. Doing so requires two steps, and while one can easily pipe these two steps together, I have broken them out into two distinct steps below for improved clarity.\n",
    "\n",
    "The first step is to use the SAMtools mpileup command to calculate the genotype likelihoods supported by the aligned reads in our sample.\n",
    "\n",
    "```bash\n",
    "Usage: samtools mpileup [options] in1.bam [in2.bam [...]]\n",
    "\n",
    "Input options:\n",
    "\n",
    "       -6           assume the quality is in the Illumina-1.3+ encoding\n",
    "       -A           count anomalous read pairs\n",
    "       -B           disable BAQ computation\n",
    "       -b FILE      list of input BAM filenames, one per line [null]\n",
    "       -C INT       parameter for adjusting mapQ; 0 to disable [0]\n",
    "       -d INT       max per-BAM depth to avoid excessive memory usage [250]\n",
    "       -E           recalculate extended BAQ on the fly thus ignoring existing BQs\n",
    "       -f FILE      faidx indexed reference sequence file [null]\n",
    "       -G FILE      exclude read groups listed in FILE [null]\n",
    "       -l FILE      list of positions (chr pos) or regions (BED) [null]\n",
    "       -M INT       cap mapping quality at INT [60]\n",
    "       -r STR       region in which pileup is generated [null]\n",
    "       -R           ignore RG tags\n",
    "       -q INT       skip alignments with mapQ smaller than INT [0]\n",
    "       -Q INT       skip bases with baseQ/BAQ smaller than INT [13]\n",
    "       --rf INT     required flags: skip reads with mask bits unset []\n",
    "       --ff INT     filter flags: skip reads with mask bits set []\n",
    "\n",
    "Output options:\n",
    "\n",
    "       -D           output per-sample DP in BCF (require -g/-u)\n",
    "       -g           generate BCF output (genotype likelihoods)\n",
    "       -O           output base positions on reads (disabled by -g/-u)\n",
    "       -s           output mapping quality (disabled by -g/-u)\n",
    "       -S           output per-sample strand bias P-value in BCF (require -g/-u)\n",
    "       -u           generate uncompress BCF output\n",
    "\n",
    "SNP/INDEL genotype likelihoods options (effective with `-g' or `-u'):\n",
    "\n",
    "       -e INT       Phred-scaled gap extension seq error probability [20]\n",
    "       -F FLOAT     minimum fraction of gapped reads for candidates [0.002]\n",
    "       -h INT       coefficient for homopolymer errors [100]\n",
    "       -I           do not perform indel calling\n",
    "       -L INT       max per-sample depth for INDEL calling [250]\n",
    "       -m INT       minimum gapped reads for indel candidates [1]\n",
    "       -o INT       Phred-scaled gap open sequencing error probability [40]\n",
    "       -p           apply -m and -F per-sample to increase sensitivity\n",
    "       -P STR       comma separated list of platforms for indels [all]\n",
    "\n",
    "Notes: Assuming diploid individuals\n",
    "```\n",
    "\n",
    "Thus we can run the following command to output the **genotype likelihood** for the possible variant sites:\n",
    "```bash\n",
    "samtools mpileup -g -f genomes/NC_008253.fna alignments/sim_reads_aligned.sorted.bam > variants/sim_variants.bcf\n",
    "```\n",
    "\n",
    "The `samtools mpileup` command automatically scans every position for each aligned read, computes **all the possible genotypes supported by the reads**, and then computes the **probability for each of these genotypes**.\n",
    "\n",
    "In our case, SAMtools scans the first 1000 bases in the E. coli genome, and then weighs the collective evidence of all reads to infer the true genotype. For example, *position 42 (reference = G) has 24 reads with a G and two reads with a T (total read depth = 26)*. In this case, SAMtools concludes with high probability that the sample has a **genotype of G**, and that **T reads are likely due to sequencing errors**.\n",
    "\n",
    "By contrast, *position 736 (reference = T) has 2 reads with a C and 66 reads with a G (total read depth = 68)*. In this case, SAMtools concludes with high probability that the sample has a genotype of G. For each position assayed, SAMtools computes all the possible genotypes, and then outputs all the results in the **binary call format (BCF)**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second step is to use `bcftools`:\n",
    "```bash\n",
    "bcftools view -c -v variants/sim_variants.bcf > variants/sim_variants.vcf\n",
    "```\n",
    "\n",
    "The `bcftools view` command uses the genotype likelihoods generated from the previous step to call SNPs and indels, and outputs the all identified variants in the **variant call format (VCF)**, which are widely used to represent genomic variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Understanding the VCF Format\n",
    "\n",
    "Similar to SAM,  VCF files contain **header lines**, which begin with # or ## symbols, and **data lines**, which do not begin with a # symbol.\n",
    "\n",
    "The header contains important information, such as the name of the program which generated the file, the VCF format version number, the reference genome name, and **information regarding individual columns within the file**.\n",
    "\n",
    "Each data line is required 8 mandatory columns:\n",
    "- **CHROM** reference, \n",
    "- **POS**ition within the chromosome, \n",
    "- **ID** of the site,\n",
    "- **REF**erence allele, \n",
    "- **ALT**ernative allele, \n",
    "- **QUAL**ilities,\n",
    "- **FILTER** or not,\n",
    "- extra **INFO**rmation\n",
    "\n",
    "\n",
    "Directly after these eight is a FORMAT column used to describe the format for all subsequent sample columns. VCF can contain information regarding multiple samples, and each sample gets its own column. \n",
    "\n",
    "In our example, we have only 1-sample.\n",
    "\n",
    "We have only one variant record:\n",
    "```\n",
    "#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  alignments/sim_read.sorted.bam\n",
    "NC_008253.1   736  .   T   G,C   33   .    DP=76;VDB=1.369620e-01;AF1=1;AC1=2;DP4=0,0,0,69;MQ=36;FQ=-229   PL  66,202,0,68,170,62\n",
    "```\n",
    "\n",
    "Within this example VCF file, the FORMAT column is set to \"PL\". If you search within the header, you can find that PL is defined as:\n",
    "```\n",
    "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">\n",
    "```\n",
    "\n",
    "Based on the reference and the alternative sequences, SAMtools automatically calculates the following genotypes, in this exact order:\n",
    "\n",
    "| Genotype\t| Phred Scaled Score |\tLikelihood p-value |\n",
    "| --- | --- | --- |\n",
    "| TT |\t64\t| $10^{-64/10}$ = 3.981e-07 |\n",
    "| TG | 178\t| $10^{-178/10}$ = 2.512e-18 |\n",
    "| GG | 0 | 1.0 |\n",
    "| TC | 66 |\t$10^{-66/10}$ = 2.512e-07 |\n",
    "| GC | 147\t| $10^{-147/10}$ = 1.995e-15 |\n",
    "| CC | 60 |\t$10^{-60/10}$ = 1.000e-06 |\n",
    "\n",
    "[Here](https://samtools.github.io/hts-specs/VCFv4.2.pdf) is the details for a VCF file.\n",
    "\n",
    "### <font color=\"red\">$\\S$Exercise 2: Variant Call Format (VCFs)</font>\n",
    "\n",
    "Try your best to interpret the follwowing \"*.vcf\" files:\n",
    "```\n",
    "#CHROM    POS    ID    REF    ALT    QUAL    FILTER    INFO    FORMAT    sample1    sample2    sample3\n",
    "3    74393    .    G    T     999    .    DP=31;AF1=0.7002;AC1=4;DP4=4,0,22,2    ;... GT:PL:DP:GQ    1/1:181, 57, 0:19:57    1/1:90,15,0:5:16    0/0:0,12,85:4:7\n",
    "```\n",
    "\n",
    "See H, Li. Bioinformatics, 27(21): 2987-2993 (2011) for details of likelihood and population genetics calculation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Visualing the mapping of reads\n",
    "\n",
    "We can use `samtools tview` for viewing the mapping results of simulated reads against the reference genome:\n",
    "```bash\n",
    "samtools tview alignments/sim_reads_aligned.sorted.bam genomes/NC_008253.fna\n",
    "```\n",
    "![](../images/tview.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. References\n",
    "\n",
    "- Cock PJ et al. [The Sanger FASTQ file format for sequences with quality scores, and the Solexa/Illumina FASTQ variants](https://www.ncbi.nlm.nih.gov/pubmed/20015970). Nucleic Acids Res. 2010 Apr;38(6)\n",
    "- Danecek P. et al. [The variant call format and VCFtools](https://www.ncbi.nlm.nih.gov/pubmed/21653522). Bioinformatics. 2011 Aug 1;27(15):2156-8.\n",
    "- Li H. et al. [The Sequence Alignment/Map format and SAMtools](https://www.ncbi.nlm.nih.gov/pubmed/19505943). Bioinformatics. 2009 Aug 15;25(16):2078-9.\n",
    "- Li H. et al. [Fast and accurate short read alignment with Burrows-Wheeler transform](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2705234). Bioinformatics. 2009; 25:1754-1760.\n",
    "- Li H. et al. [Fast and accurate long-read alignment with Burrows-Wheeler Transform](https://www.ncbi.nlm.nih.gov/pubmed/20080505). Bioinformatics. 2010;26:589-595."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Exercises\n",
    "\n",
    "1. Try organizing the above process into an analysis pipeline.\n",
    "2. Try a larger set of different length of paired-end reads, conduct the same analysis.\n",
    "3. Try a longer reads, and compare the performance of `bwa-backtrack`, `bwa-sw` and `bwa-mem` alignment algorithms.\n",
    "4. Try a real data on http://cbb.sjtu.edu.cn/course/advanced/lab2.pdf."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
